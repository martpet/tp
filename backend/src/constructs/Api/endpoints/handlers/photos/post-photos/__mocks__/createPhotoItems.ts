export const createPhotoItems = vi
  .fn()
  .mockName('createPhotoItems')
  .mockResolvedValue(undefined);
