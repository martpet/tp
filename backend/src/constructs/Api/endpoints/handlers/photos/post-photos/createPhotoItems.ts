import {
  BatchWriteCommand,
  DynamoDBClient,
  DynamoDBDocument,
  PhotosTableItem,
  photosTableOptions,
  PostPhotosRequest,
  unmarshall,
} from 'lambda-layer';

const ddbClient = new DynamoDBClient({});
const ddbDocClient = DynamoDBDocument.from(ddbClient);

type Props = {
  requestData: PostPhotosRequest;
  sub: string;
};

export const createPhotoItems = async ({ requestData, sub }: Props) => {
  const { tableName } = photosTableOptions;

  async function batchWrite(photoItems: PhotosTableItem[]) {
    const batchWriteCommand = new BatchWriteCommand({
      RequestItems: { [tableName]: photoItems.map((Item) => ({ PutRequest: { Item } })) },
    });
    const { UnprocessedItems } = await ddbDocClient.send(batchWriteCommand);

    if (UnprocessedItems) {
      const nextItems: PhotosTableItem[] = [];
      UnprocessedItems[tableName]?.forEach(({ PutRequest }) => {
        if (PutRequest?.Item) {
          nextItems.push(unmarshall(PutRequest.Item) as PhotosTableItem);
        }
      });
      if (nextItems.length) {
        await batchWrite(nextItems);
      }
    }
  }

  const batchSize = 25;
  const requests = [];

  for (let i = 0; i < requestData.length; i += batchSize) {
    const chunk = requestData.slice(i, i + batchSize);
    const photoItems = chunk.map((item) => ({
      ...item,
      userId: sub,
      createdAt: Date.now(),
    }));
    requests.push(batchWrite(photoItems));
  }

  await Promise.all(requests);
};
