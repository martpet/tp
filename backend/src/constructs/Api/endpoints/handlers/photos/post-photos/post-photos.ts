import {
  APIGatewayProxyHandlerV2,
  ApiRouteHeaders,
  errorResponse,
  findExistingFingerprints,
  getIdTokenPayload,
  PostPhotosRequest,
  PostPhotosResponse,
  StatusCodes,
} from 'lambda-layer';

import { createPhotoItems } from './createPhotoItems';

export const handler: APIGatewayProxyHandlerV2<PostPhotosResponse> = async (event) => {
  const { authorization } = event.headers as ApiRouteHeaders<'/settings'>;

  if (!authorization) {
    return errorResponse('f31e4f94d3');
  }

  if (!event.body) {
    return errorResponse('24d59e00da', { statusCode: StatusCodes.BAD_REQUEST });
  }

  let requestData;

  try {
    requestData = JSON.parse(event.body) as PostPhotosRequest;
  } catch (error) {
    return errorResponse('d45458a241', { statusCode: StatusCodes.BAD_REQUEST, error });
  }

  const existingFingerprintsInDb = await findExistingFingerprints(
    requestData.map(({ fingerprint }) => fingerprint)
  );

  if (existingFingerprintsInDb.length) {
    return errorResponse('9082bdd5f9', {
      statusCode: StatusCodes.FORBIDDEN,
      description: 'Cannot replace existing item',
    });
  }

  const { sub } = await getIdTokenPayload(authorization);
  return createPhotoItems({ requestData, sub });
};
