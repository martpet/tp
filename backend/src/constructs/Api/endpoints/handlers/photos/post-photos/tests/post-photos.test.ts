import {
  findExistingFingerprints,
  itCalls,
  itGetsIdToken,
  itHasJsonBody,
  itResolves,
  itResolvesWithError,
} from '~/constructs/Api/utils';

import { createPhotoItems } from '../createPhotoItems';
import { handler } from '../post-photos';

vi.mock('~/constructs/Api/utils/errorResponse');
vi.mock('~/constructs/Api/utils/getIdTokenPayload');
vi.mock('~/constructs/Api/utils/XfindExistingFingerprints');
vi.mock('../createPhotoItems');

vi.mocked(findExistingFingerprints).mockResolvedValue([]);

const args = [
  {
    headers: { authorization: 'dummyAuthorizationHeader' },
    body: JSON.stringify([{ fingerprint: 'dummyFingerprint1' }]),
  },
] as unknown as Parameters<typeof handler>;

describe('post-photos', () => {
  itHasJsonBody(handler, args);
  itGetsIdToken(handler, args);
  itCalls(findExistingFingerprints, handler, args);
  itCalls(createPhotoItems, handler, args);
  itResolves(handler, args);

  describe('when "findExistingFingerprints" has results', () => {
    beforeEach(() => {
      vi.mocked(findExistingFingerprints).mockResolvedValueOnce([
        'dummyExistingFingerprint',
      ]);
    });
    itResolvesWithError(handler, args);
  });
});
