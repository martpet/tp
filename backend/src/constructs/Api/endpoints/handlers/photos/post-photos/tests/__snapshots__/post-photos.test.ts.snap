// Vitest Snapshot v1

exports[`post-photos > calls "createPhotoItems" with correct args 1`] = `
[
  [
    {
      "requestData": [
        {
          "fingerprint": "dummyFingerprint1",
        },
      ],
      "sub": "dummySub",
    },
  ],
]
`;

exports[`post-photos > calls "findExistingFingerprints" with correct args 1`] = `
[
  [
    [
      "dummyFingerprint1",
    ],
  ],
]
`;

exports[`post-photos > calls "getIdTokenPayload" with correct args 1`] = `
[
  [
    "dummyAuthorizationHeader",
  ],
]
`;

exports[`post-photos > resolves with a correct value 1`] = `undefined`;

exports[`post-photos > when "authorization" header is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "f31e4f94d3",
  ],
]
`;

exports[`post-photos > when "authorization" header is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`post-photos > when "event.body" is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "24d59e00da",
    {
      "statusCode": 400,
    },
  ],
]
`;

exports[`post-photos > when "event.body" is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`post-photos > when "event.body" is not JSON > calls "errorResponse" with correct args 1`] = `
[
  [
    "d45458a241",
    {
      "error": [SyntaxError: Unexpected token o in JSON at position 1],
      "statusCode": 400,
    },
  ],
]
`;

exports[`post-photos > when "event.body" is not JSON > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`post-photos > when "findExistingFingerprints" has results > calls "errorResponse" with correct args 1`] = `
[
  [
    "9082bdd5f9",
    {
      "description": "Cannot replace existing item",
      "statusCode": 403,
    },
  ],
]
`;

exports[`post-photos > when "findExistingFingerprints" has results > resolves with a correct value 1`] = `"dummyErrorResponse"`;
